src/closed.jl

```diff
@@ -0,0 +1,54 @@
+"""
+Defines the ClosedSet type.
+
+A ClosedSet(left, right) is an interval set where both bounds are closed i.e. the range is [left, right]
+"""
+immutable ClosedSet{T}
+    left::T
+    right::T
+end
+
+function ClosedSet(left, right)
+    (I_left, I_right) = promote(left, right)
+    ClosedSet(I_left, I_right)
+end
+
+..(x, y) = ClosedSet(x, y)
+±(x, y) = ClosedSet(x - y, x + y)
+±(x::CartesianIndex, y) = map(ClosedSet, (x - y).I, (x + y).I)
+
+show(io::IO, I::ClosedSet) = print(io, I.left, "..", I.right)
+
+in(v, I::ClosedSet) = I.left <= v <= I.right
+
+isempty(A::ClosedSet) = A.left > A.right
+
+isequal(A::ClosedSet, B::ClosedSet) = (A.left == B.left && A.right == B.right) || isempty(A) && isempty(B)
+
+==(A::ClosedSet, B::ClosedSet) = isequal(A, B)
+
+function intersect(A::ClosedSet, B::ClosedSet)
+    left = max(A.left, B.left)
+    right = min(A.right, B.right)
+    ClosedSet(left, right)
+end
+
+function union{T<:AbstractFloat}(A::ClosedSet{T}, B::ClosedSet{T})
+    @assert max(A.left, B.left) <= nextfloat(min(A.right, B.right)) throw(ArgumentError("The sets are disjoint."))
```

Comment:
Rather than using `@assert`, I'd recommend the `condition || throw(...)` syntax. I'd also maybe make the error more specific, maybe something like `Cannot construct the union of disjoint sets`?